Introduce descriptors into Bionic internals

jqmp opened the pull request:
This moves a lot of Bionic's internal code to using DescriptorNodes, or "dnodes", instead of simple entity names. Currently dnodes are just wrappers for entity names, but in the future they will be able to represent more complex structures.

There should be no functional changes in this commit, except that we're updating the cache schema version.

Finally, this commit introduces a dependency on the [attrs](https://www.attrs.org/en/stable/) package. I think this is a cleaner way to create customized data classes than `namedtuple`, and I'd be interested in converting some of the classes in `datatypes.py` to use `attrs` as well.

namanjain commented on bionic/deriver.py:
```diff
@@ -181,47 +184,48 @@ def _initialize_task_state(self, task_state):
 
         dep_key_spaces_by_name = {
-            dep_name: self._key_spaces_by_entity_name[dep_name]
+            dep_name: self._key_spaces_by_dnode[
```
same here, why name and not dnode? Actually, same comment for all other maps in this PR too.

namanjain commented on bionic/cache.py:
```diff
@@ -809,7 +811,18 @@ class InvalidCacheStateError(Exception):
 
 
-CACHE_SCHEMA_VERSION = 5
+def filename_from_query(query):
```
I don't know if this is the best name since it's not really used as a file name but as stem and a dir. How about something like `encoded_descriptor_from_query`?

jqmp replied:
My thinking here is that the point is to return something that would be appropriate for a filename. In some places we use it as a full filename and in others we use it as a stem.

Maybe `valid_filename_from_query`?

namanjain replied:
Where are we using it as a full filename? I see that it is used a dir name in other places but not full filename.

jqmp replied:
Oh, I was counting a directory name as a filename.

jqmp commented on bionic/descriptors.py:
```diff
@@ -0,0 +1,72 @@
+This module defines functions and data structures relating to descriptors.
+A descriptor is intended to be a generalized entity name; it's a short string
+expression that can represent the input or output of a function in Bionic. For
+example, instead of instead of referring only to atomic entities like `raw_data`
```
thanks thanks

jqmp commented on bionic/deriver.py:
```diff
@@ -181,47 +184,48 @@ def _initialize_task_state(self, task_state):
+        entity_name = dnode.to_descriptor()
         provider = self._flow_state.get_provider(entity_name)
 
         dep_names = provider.get_dependency_names()
```
Yep, we definitely can. I have another commit that does this. I might just add it onto this PR.

jqmp commented on bionic/deriver.py:
```diff
@@ -181,47 +184,48 @@ def _initialize_task_state(self, task_state):
+        if dnode in self._task_lists_by_dnode:
             return
 
+        entity_name = dnode.to_descriptor()
```
Sorry, this line should probably be `entity_name = dnode.to_entity_name()`. In this case it's important that it be an actual entity name, since that's what `FlowState.get_provider()` expects.

jqmp commented on bionic/descriptors.py:
```diff
@@ -0,0 +1,72 @@
+
+
+@attr.s(frozen=True)
+class EntityNode:
```
Oops, I intended for it to subclass `DescriptorNode` but forgot to. I guess for now the duck typing is making everything work. (Because we only have one concrete implementation so far, I was considering not even including a base class in this PR, but I ended up re-adding it and forgetting to update `EntityNode`.)

I had been going back and forth about whether to include a `DescriptorNode` class at all, since `EntityNode` is the only concrete implementation so far. I guess in this case the duck typing makes everything work,

jqmp commented on bionic/flow.py:
```diff
@@ -987,7 +988,9 @@ def get(self, name, collection=None, fmt=None, mode=object):
 
-        result_group = self._deriver.derive(name)
+        dnode = DescriptorNode.from_descriptor(name)
+        result_group = self._deriver.derive(dnode)
```
Whoa, good catch! I'm a little surprised we don't have any tests for `export` anymore.

jqmp commented on bionic/provider.py:
```diff
@@ -275,6 +276,7 @@ def __init__(self, name, protocol, doc):
         )
 
         self.name = name
```
Maybe not! A lot of the provider stuff is addressed in the next stage, but not this particular field. I'll check.

jqmp commented:
I added another commit here that refactors the providers as well. This addresses some of your comments; I'll push another commit later that addresses the rest (including the typos, problem with `export`, etc.)

namanjain commented on bionic/deriver.py:
```diff
@@ -80,8 +82,8 @@ def should_include_entity_name(name):
-        for entity_name, tasks in (
-                self._task_lists_by_entity_name.items()):
+        for dnode, tasks in self._task_lists_by_dnode.items():
+            entity_name = dnode.to_descriptor()
```
`to_descriptor` -> `to_entity_name()`

jqmp commented on bionic/deriver.py:
```diff
@@ -181,47 +184,48 @@ def _initialize_task_state(self, task_state):
         # All names in this task should point to the same provider.
         task_state.provider, = set(
-            self._flow_state.get_provider(task_key.entity_name)
+            self._flow_state.get_provider(task_key.dnode.to_entity_name())
```
This is related to the following comment but I wanted to expand a bit here: I've left `FlowState.get_provider` taking an entity name, not a generic descriptor, because a flow state is a mapping from entities (not descriptors) to providers. In the future the interface will probably change, but I'm not sure how yet.
